From e60d689361b27004fa6fff1cdb26d895b1be72f7 Mon Sep 17 00:00:00 2001
From: Alexander Koskovich <akoskovich@pm.me>
Date: Sun, 04 Oct 2026 15:41:30 -0400
Subject: [PATCH] Revert "ipc: Handle DSP metadata for non-tunnel streams"

This reverts commit 009ffd26adc3c167ee780657fec6ccf4e2042852.

Breaks libqc2audio_hwaudiocodec, could not find an SM8450/SM8475 device
with blobs from the tag that introduced this change or beyond.

Test: Play AAC through c2.qti.aac.hw.decoder, no SIGSEGV
Change-Id: I615dd50645c25c72c16000b48fd44b7c5747858a
---

diff --git a/ipc/HwBinders/interfaces/pal/1.0/IPALCallback.hal b/ipc/HwBinders/interfaces/pal/1.0/IPALCallback.hal
index 1d84415..0dac77b 100644
--- a/ipc/HwBinders/interfaces/pal/1.0/IPALCallback.hal
+++ b/ipc/HwBinders/interfaces/pal/1.0/IPALCallback.hal
@@ -25,10 +25,6 @@
  * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
  * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
  * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
- *
- * Changes from Qualcomm Innovation Center are provided under the following license:
- * Copyright (c) 2025 Qualcomm Innovation Center, Inc. All rights reserved.
- * SPDX-License-Identifier: BSD-3-Clause-Clear
  */
 
 package vendor.qti.hardware.pal@1.0;
@@ -38,9 +34,9 @@
     event_callback (PalStreamHandle streamHandle, uint32_t event_id,
                     uint32_t event_data_size, vec<uint8_t> event_data,
                     uint64_t cookie) generates(int32_t ret);
-    oneway event_callback_rw_done (PalStreamHandle streamHandle, uint32_t event_id,
+    event_callback_rw_done (PalStreamHandle streamHandle, uint32_t event_id,
                             uint32_t event_data_size,
-                            vec<PalCallbackBuffer> rw_done_payload,
-                            uint64_t cookie);
+                            vec<PalEventReadWriteDonePayload> rw_done_payload,
+                            uint64_t cookie) generates(int32_t ret);
 };
 
diff --git a/ipc/HwBinders/interfaces/pal/1.0/types.hal b/ipc/HwBinders/interfaces/pal/1.0/types.hal
index e17d66e..1f88e32 100644
--- a/ipc/HwBinders/interfaces/pal/1.0/types.hal
+++ b/ipc/HwBinders/interfaces/pal/1.0/types.hal
@@ -25,10 +25,6 @@
  * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
  * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
  * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
- *
- * Changes from Qualcomm Innovation Center are provided under the following license:
- * Copyright (c) 2025 Qualcomm Innovation Center, Inc. All rights reserved.
- * SPDX-License-Identifier: BSD-3-Clause-Clear
  */
 
 package vendor.qti.hardware.pal@1.0;
@@ -270,28 +266,14 @@
 
 /** PAL buffer structure used for reading/writing buffers from/to the stream */
 struct PalBuffer {
-    vec<uint8_t> buffer;                // buffer pointer
+    vec<uint8_t> buffer;                //  buffer pointer
     uint32_t size;                      // number of bytes
     uint32_t offset;                    // offset in buffer from where valid byte starts
-    TimeSpec timeStamp;                 // timestamp
-    uint32_t flags;                     // stream flags
-    PalExternAllocBuffInfo alloc_info;  // holds info for extern buff
-    uint64_t frame_index;               // frame index of the buffer
-};
-
-struct PalCallbackBufferInfo {
-    uint64_t frame_index;       // frame index of the buffer
-    uint32_t sample_rate;       // updated sample rate
-    uint32_t bit_width;         // updated bit width
-    uint16_t channel_count;     // updated channel count
-};
-
-struct PalCallbackBuffer {
-    vec<uint8_t> buffer;                // buffer pointer
-    uint32_t size;                      // filled length of the buffer
-    TimeSpec timeStamp;                 // timestamp
-    uint32_t status;                    // status of payload
-    PalCallbackBufferInfo cbBufInfo;    // buffer info
+    TimeSpec timeStamp;                 // timestmap
+    uint32_t flags;             // meta data flags
+    uint32_t metadataSz;
+    vec<uint8_t> metadata;
+    PalExternAllocBuffInfo alloc_info; /**< holds info for extern buff */
 };
 
 struct PalChannelVolKv {
diff --git a/ipc/HwBinders/interfaces/pal/current.txt b/ipc/HwBinders/interfaces/pal/current.txt
index f723476..698a1bc 100644
--- a/ipc/HwBinders/interfaces/pal/current.txt
+++ b/ipc/HwBinders/interfaces/pal/current.txt
@@ -1,5 +1,5 @@
 # Hash for vendor.qti.hardware.pal@1.0 package
-30f77bb61842e192ee3d92187ca69ff9b95d27a98d9a1aaee9ac5335f29159da vendor.qti.hardware.pal@1.0::types
+d2952e2076bed0f206a84e87c2ef242d68ed1f3bb8bf8a2a42a109be974b99d8 vendor.qti.hardware.pal@1.0::types
 490e78d428acdd280e198ae7071ffee6abfe790aff0f649653a619cc8ee5cf26 vendor.qti.hardware.pal@1.0::IPAL
-78f4a508b2a1dc442234f969f3e3251d17ddf665bbe8e91a670977e54ab023c7 vendor.qti.hardware.pal@1.0::IPALCallback
+d6ae25f7077995036a155000e292422955e3c5515887d76947625337c7f8b9b6 vendor.qti.hardware.pal@1.0::IPALCallback
 
diff --git a/ipc/HwBinders/pal_ipc_client/inc/PalCallback.h b/ipc/HwBinders/pal_ipc_client/inc/PalCallback.h
index b7a7773..cf5bb91 100644
--- a/ipc/HwBinders/pal_ipc_client/inc/PalCallback.h
+++ b/ipc/HwBinders/pal_ipc_client/inc/PalCallback.h
@@ -25,10 +25,6 @@
  * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
  * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
  * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
- *
- * Changes from Qualcomm Innovation Center are provided under the following license:
- * Copyright (c) 2025 Qualcomm Innovation Center, Inc. All rights reserved.
- * SPDX-License-Identifier: BSD-3-Clause-Clear
  */
 
 #pragma once
@@ -101,9 +97,9 @@
                                uint32_t event_id, uint32_t event_data_size,
                                const hidl_vec<uint8_t>& event_data,
                                uint64_t cookie) override;
-    Return<void> event_callback_rw_done(uint64_t stream_handle,
+    Return<int32_t> event_callback_rw_done(uint64_t stream_handle,
                                uint32_t event_id, uint32_t event_data_size,
-                               const hidl_vec<PalCallbackBuffer>& event_data,
+                               const hidl_vec<PalEventReadWriteDonePayload>& event_data,
                                uint64_t cookie) override;
 
     PalCallback(pal_stream_callback callBack)
diff --git a/ipc/HwBinders/pal_ipc_client/src/pal_client_wrapper.cpp b/ipc/HwBinders/pal_ipc_client/src/pal_client_wrapper.cpp
index ff03839..e8606c5 100644
--- a/ipc/HwBinders/pal_ipc_client/src/pal_client_wrapper.cpp
+++ b/ipc/HwBinders/pal_ipc_client/src/pal_client_wrapper.cpp
@@ -26,10 +26,6 @@
  * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
  * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
  * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
- *
- * Changes from Qualcomm Innovation Center are provided under the following license:
- * Copyright (c) 2025 Qualcomm Innovation Center, Inc. All rights reserved.
- * SPDX-License-Identifier: BSD-3-Clause-Clear
  */
 
 #define LOG_TAG "pal_client_wrapper"
@@ -118,48 +114,91 @@
     return int32_t {};
 }
 
-Return<void> PalCallback::event_callback_rw_done(uint64_t strm_handle,
+Return<int32_t> PalCallback::event_callback_rw_done(uint64_t strm_handle,
                                  uint32_t event_id,
                                  uint32_t event_data_size,
-                                 const hidl_vec<PalCallbackBuffer>& event_data,
+                                 const hidl_vec<PalEventReadWriteDonePayload>& event_data,
                                  uint64_t cookie) {
-    const PalCallbackBuffer *rwDonePayloadHidl = event_data.data();
+    struct pal_event_read_write_done_payload *rw_done_payload;
+    struct pal_buffer *buffer = nullptr;
+    uint32_t *ev_data = NULL;
+    const native_handle *allochandle = nullptr;
+    const PalEventReadWriteDonePayload *rwDonePayloadHidl = event_data.data();
+    PalBuffer *bufferHidl;
+
     ALOGV("%s called \n", __func__);
-
-    auto cbBuffer = std::make_unique<pal_callback_buffer>();
-    if (!cbBuffer) {
-        ALOGE("%s:%d Failed to allocate memory for callback payload", __func__, __LINE__);
-        return Void();
+    rw_done_payload = (struct pal_event_read_write_done_payload *)
+                          calloc(1, sizeof(pal_event_read_write_done_payload));
+    if (!rw_done_payload) {
+        ALOGE("%s:%d Failed to allocate memory to rw_done_payload", __func__, __LINE__);
+        goto exit;
     }
 
-    cbBuffer->size = rwDonePayloadHidl->size;
-    std::vector<uint8_t> buffData = {};
-    if (rwDonePayloadHidl->buffer.size() == cbBuffer->size) {
-        buffData.resize(cbBuffer->size);
-        memcpy(buffData.data(), rwDonePayloadHidl->buffer.data(), cbBuffer->size);
-        cbBuffer->buffer = buffData.data();
+    ev_data = (uint32_t *)rw_done_payload;
+    rw_done_payload->tag = rwDonePayloadHidl->tag;
+    rw_done_payload->status = rwDonePayloadHidl->status;
+    rw_done_payload->md_status = rwDonePayloadHidl->md_status;
+
+    buffer = &rw_done_payload->buff;
+
+    buffer->size = rwDonePayloadHidl->buff.size;
+    if (rwDonePayloadHidl->buff.buffer.size() == buffer->size) {
+        buffer->buffer = (uint8_t *)calloc(1, buffer->size);
+        if (!buffer->buffer) {
+            ALOGE("%s:%d Failed to allocate memory to buffer", __func__, __LINE__);
+            goto exit;
+        }
+        memcpy(buffer->buffer, rwDonePayloadHidl->buff.buffer.data(),
+               buffer->size);
     }
 
-    auto bufTimeSpec = std::make_unique<timespec>();
-    if (!bufTimeSpec) {
-        ALOGE("%s:%d Failed to allocate memory for timespec", __func__, __LINE__);
-        return Void();
+    buffer->offset = rwDonePayloadHidl->buff.offset;
+    buffer->ts = (struct timespec *) calloc(1, sizeof(struct timespec));
+    if (!buffer->ts) {
+        ALOGE("%s:%d Failed to allocate memory to buffer->ts", __func__, __LINE__);
+        goto exit;
     }
-    bufTimeSpec->tv_sec = rwDonePayloadHidl->timeStamp.tvSec;
-    bufTimeSpec->tv_nsec = rwDonePayloadHidl->timeStamp.tvNSec;
-    cbBuffer->ts = (timespec *) bufTimeSpec.get();
-    cbBuffer->status = rwDonePayloadHidl->status;
-    cbBuffer->cb_buf_info.frame_index = rwDonePayloadHidl->cbBufInfo.frame_index;
-    cbBuffer->cb_buf_info.sample_rate = rwDonePayloadHidl->cbBufInfo.sample_rate;
-    cbBuffer->cb_buf_info.bit_width = rwDonePayloadHidl->cbBufInfo.bit_width;
-    cbBuffer->cb_buf_info.channel_count = rwDonePayloadHidl->cbBufInfo.channel_count;
-    ALOGV("%s:%d Bufsize %d  ret bufSize %zu", __func__, __LINE__,
-                rwDonePayloadHidl->size, cbBuffer->size);
-    ALOGV("event_payload_size %d", event_data_size);
-    this->cb((pal_stream_handle_t *)strm_handle, event_id, (uint32_t *)cbBuffer.get(),
-                                    event_data_size, cookie);
 
-    return Void();
+    buffer->ts->tv_sec = rwDonePayloadHidl->buff.timeStamp.tvSec;
+    buffer->ts->tv_nsec = rwDonePayloadHidl->buff.timeStamp.tvNSec;
+    buffer->flags = rwDonePayloadHidl->buff.flags;
+    if (rwDonePayloadHidl->buff.metadataSz) {
+        buffer->metadata_size = rwDonePayloadHidl->buff.metadataSz;
+        buffer->metadata = (uint8_t *)calloc(1, buffer->metadata_size);
+        if (!buffer->metadata) {
+            ALOGE("%s:%d Failed to allocate memory to buffer->metadata",
+                __func__, __LINE__);
+            goto exit;
+        }
+
+        ALOGV("metadatasize %zu \n", buffer->metadata_size);
+        memcpy(buffer->metadata, rwDonePayloadHidl->buff.metadata.data(),
+               buffer->metadata_size);
+    }
+
+    allochandle = rwDonePayloadHidl->buff.alloc_info.alloc_handle.handle();
+
+    buffer->alloc_info.alloc_handle = allochandle->data[1];
+    buffer->alloc_info.alloc_size = rwDonePayloadHidl->buff.alloc_info.alloc_size;
+    buffer->alloc_info.offset = rwDonePayloadHidl->buff.alloc_info.offset;
+    ALOGV("%s:%d Bufsize %d  ret bufSize %zu", __func__, __LINE__, rwDonePayloadHidl->buff.size, buffer->size);
+    ALOGV("event_payload_size %d alloc_handle %d", event_data_size, allochandle->data[1]);
+    ALOGV("alloc size %d alloc_size ret %d", rwDonePayloadHidl->buff.alloc_info.alloc_size,buffer->alloc_info.alloc_size);
+    this->cb((pal_stream_handle_t *)strm_handle, event_id, ev_data, event_data_size,
+                                    cookie);
+
+exit:
+    if (buffer) {
+        if (buffer->metadata)
+            free(buffer->metadata);
+        if (buffer->buffer)
+            free(buffer->buffer);
+        if (buffer->ts)
+            free(buffer->ts);
+    }
+    if (rw_done_payload)
+        free(rw_done_payload);
+    return int32_t {};
 }
 
 int32_t pal_stream_open(struct pal_stream_attributes *attr,
@@ -412,13 +451,13 @@
 {
     int ret = -EINVAL;
 
-    if (!stream_handle)
-        return ret;
+    if (stream_handle == NULL)
+       goto done;
 
     if (!pal_server_died) {
         ALOGV("%s:%d hndl %p",__func__, __LINE__, stream_handle );
         android::sp<IPAL> pal_client = get_pal_server();
-        if (!pal_client)
+        if (pal_client == nullptr)
             return ret;
 
         hidl_vec<PalBuffer> buf_hidl;
@@ -437,24 +476,28 @@
         palBuff->offset = buf->offset;
         palBuff->buffer.resize(buf->size);
         palBuff->flags = buf->flags;
-        palBuff->frame_index = buf->frame_index;
         if (buf->ts) {
              palBuff->timeStamp.tvSec = buf->ts->tv_sec;
              palBuff->timeStamp.tvNSec = buf->ts->tv_nsec;
         }
         if (buf->size && buf->buffer)
             memcpy(palBuff->buffer.data(), buf->buffer, buf->size);
-         palBuff->alloc_info.alloc_handle =
-                 hidl_memory("arpal_alloc_handle", hidl_handle(allocHidlHandle),
-                              buf->alloc_info.alloc_size);
-         ALOGV("%s:%d alloc handle %d sending %d",__func__,__LINE__,
-                     buf->alloc_info.alloc_handle, allocHidlHandle->data[0]);
+        if ((buf->metadata_size > 0) && buf->metadata) {
+            palBuff->metadataSz = buf->metadata_size;
+            palBuff->metadata.resize(buf->metadata_size);
+            memcpy(palBuff->metadata.data(),
+                   buf->metadata, buf->metadata_size);
+         }
+         palBuff->alloc_info.alloc_handle = hidl_memory("arpal_alloc_handle", hidl_handle(allocHidlHandle),
+                                                         buf->alloc_info.alloc_size);
+         ALOGV("%s:%d alloc handle %d sending %d",__func__,__LINE__, buf->alloc_info.alloc_handle, allocHidlHandle->data[0]);
          palBuff->alloc_info.alloc_size = buf->alloc_info.alloc_size;
          palBuff->alloc_info.offset = buf->alloc_info.offset;
          ret = pal_client->ipc_pal_stream_write((PalStreamHandle)stream_handle, buf_hidl);
          if (allocHidlHandle)
              native_handle_delete(allocHidlHandle);
     }
+done:
     return ret;
 }
 
@@ -462,12 +505,11 @@
 {
     int ret = -EINVAL;
 
-    if (!stream_handle)
-        return ret;
-
+    if (stream_handle == NULL)
+       goto done;
     if (!pal_server_died) {
         android::sp<IPAL> pal_client = get_pal_server();
-        if (!pal_client)
+        if (pal_client == nullptr)
             return ret;
 
         hidl_vec<PalBuffer> buf_hidl;
@@ -484,6 +526,7 @@
 
         palBuff->size = buf->size;
         palBuff->offset = buf->offset;
+        palBuff->metadataSz = buf->metadata_size;
         palBuff->alloc_info.alloc_handle = hidl_memory("arpal_alloc_handle", hidl_handle(allocHidlHandle),
                                                          buf->alloc_info.alloc_size);
         palBuff->alloc_info.alloc_size = buf->alloc_info.alloc_size;
@@ -501,6 +544,10 @@
                                      ret_buf_hidl.data()->size, buf->size);
                               ret_ = -ENOMEM;
                            } else {
+                              if ((buf->metadata_size > 0) && buf->metadata)
+                                  memcpy(buf->metadata,
+                                         ret_buf_hidl.data()->metadata.data(),
+                                         ret_buf_hidl.data()->metadataSz);
                               if (buf->ts) {
                                   buf->ts->tv_sec =
                                        ret_buf_hidl.data()->timeStamp.tvSec;
@@ -521,6 +568,7 @@
         if (allocHidlHandle)
             native_handle_delete(allocHidlHandle);
     }
+done:
     return ret;
 }
 
diff --git a/ipc/HwBinders/pal_ipc_server/src/pal_server_wrapper.cpp b/ipc/HwBinders/pal_ipc_server/src/pal_server_wrapper.cpp
index 38a35bd..0426ea4 100644
--- a/ipc/HwBinders/pal_ipc_server/src/pal_server_wrapper.cpp
+++ b/ipc/HwBinders/pal_ipc_server/src/pal_server_wrapper.cpp
@@ -27,14 +27,13 @@
  * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
  * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  *
- * Changes from Qualcomm Innovation Center are provided under the following license:
- * Copyright (c) 2022-2025 Qualcomm Innovation Center, Inc. All rights reserved.
+ * Changes from Qualcomm Innovation Center, Inc. are provided under the following license:
+ * Copyright (c) 2022-2024 Qualcomm Innovation Center, Inc. All rights reserved.
  * SPDX-License-Identifier: BSD-3-Clause-Clear
  */
 
 #define LOG_TAG "pal_server_wrapper"
 #include "inc/pal_server_wrapper.h"
-#include "MetadataParser.h"
 #include <hwbinder/IPCThreadState.h>
 
 #define MAX_CACHE_SIZE 64
@@ -43,49 +42,6 @@
 using android::hardware::hidl_handle;
 using android::hardware::hidl_memory;
 
-// map<FD, map<offset, input_frame_id>>
-std::map<int, std::map<uint32_t, uint64_t>> gInputsPendingAck;
-std::mutex gInputsPendingAckLock;
-
-void addToPendingInputs(int fd, uint32_t offset, uint32_t ip_frame_id) {
-    ALOGV("%s: fd %d, offset %u", __func__, fd, offset);
-    std::lock_guard<std::mutex> pendingAcksLock(gInputsPendingAckLock);
-    auto itFd = gInputsPendingAck.find(fd);
-    if (itFd != gInputsPendingAck.end()) {
-        gInputsPendingAck[fd][offset] = ip_frame_id;
-        ALOGV("%s: added offset %u and frame id %u", __func__, offset,
-               (uint32_t)gInputsPendingAck[fd][offset]);
-    } else {
-        //create new map<offset,input_buffer_index> and add to FD map
-        ALOGV("%s: added map for fd %lu", __func__, (unsigned long)fd);
-        gInputsPendingAck.insert(std::make_pair(fd, std::map<uint32_t, uint64_t>()));
-        ALOGV("%s: added frame id %lu for fd %d offset %u", __func__,
-                    (unsigned long)ip_frame_id,  fd, (unsigned int)offset);
-        gInputsPendingAck[fd].insert(std::make_pair(offset, ip_frame_id));
-    }
-}
-
-int getInputBufferIndex(int fd, uint32_t offset, uint64_t &buf_index) {
-    int status = 0;
-
-    std::lock_guard<std::mutex> pendingAcksLock(gInputsPendingAckLock);
-    ALOGV("%s: fd %d, offset %u", __func__, fd, offset);
-    std::map<int, std::map<uint32_t, uint64_t>>::iterator itFd = gInputsPendingAck.find(fd);
-    if (itFd != gInputsPendingAck.end()) {
-        std::map<uint32_t, uint64_t> offsetToFrameIdxMap = itFd->second;
-        auto itOffsetFrameIdxPair = offsetToFrameIdxMap.find(offset);
-        if (itOffsetFrameIdxPair != offsetToFrameIdxMap.end()){
-            buf_index = itOffsetFrameIdxPair->second;
-            ALOGV("%s ip_frame_id=%lu", __func__, (unsigned long)buf_index);
-        } else {
-            status = -EINVAL;
-            ALOGE("%s: Entry doesn't exist for FD 0x%x and offset 0x%x",
-                    __func__, fd, offset);
-        }
-        gInputsPendingAck.erase(itFd);
-    }
-    return status;
-}
 
 namespace vendor {
 namespace qti {
@@ -201,11 +157,17 @@
     if ((sr_clbk_dat->session_attr.type == PAL_STREAM_NON_TUNNEL) &&
           ((event_id == PAL_STREAM_CBK_EVENT_READ_DONE) ||
            (event_id == PAL_STREAM_CBK_EVENT_WRITE_READY))) {
-        hidl_vec<PalCallbackBuffer> rwDonePayloadHidl;
-        PalCallbackBuffer *rwDonePayload;
+        hidl_vec<PalEventReadWriteDonePayload> rwDonePayloadHidl;
+        PalEventReadWriteDonePayload *rwDonePayload;
         struct pal_event_read_write_done_payload *rw_done_payload;
         int input_fd = -1;
         int fdToBeClosed = -1;
+        native_handle_t *allocHidlHandle = nullptr;
+        allocHidlHandle = native_handle_create(1, 1);
+        if (!allocHidlHandle) {
+            ALOGE("handle allocHidlHandle is NULL");
+            return -EINVAL;
+        }
 
         rw_done_payload = (struct pal_event_read_write_done_payload *)event_data;
         /*
@@ -217,95 +179,71 @@
             std::lock_guard<std::mutex> lock(s->mActiveSessionsLock);
             for (int idx = 0; idx < s->mActiveSessions.size(); idx++) {
                 session_info session = s->mActiveSessions[idx];
-                if (session.session_handle != (uint64_t)stream_handle) {
-                    continue;
-                }
-                std::vector<std::pair<int, int>>::iterator it;
-                for (int i = 0; i < sr_clbk_dat->sharedMemFdList.size(); i++) {
-                    if (sr_clbk_dat->sharedMemFdList[i].second ==
-                            rw_done_payload->buff.alloc_info.alloc_handle) {
-                        input_fd = sr_clbk_dat->sharedMemFdList[i].first;
-                        it = (sr_clbk_dat->sharedMemFdList.begin() + i);
-                        if (it != sr_clbk_dat->sharedMemFdList.end()) {
-                            fdToBeClosed = sr_clbk_dat->sharedMemFdList[i].second;
-                            sr_clbk_dat->sharedMemFdList.erase(it);
-                            ALOGV("Removing fd [input %d - dup %d]", input_fd, fdToBeClosed);
+                if (session.session_handle == (uint64_t)stream_handle) {
+                    std::vector<std::pair<int, int>>::iterator it;
+                    for (int i = 0; i < sr_clbk_dat->sharedMemFdList.size(); i++) {
+                        if (sr_clbk_dat->sharedMemFdList[i].second ==
+                                rw_done_payload->buff.alloc_info.alloc_handle) {
+                            input_fd = sr_clbk_dat->sharedMemFdList[i].first;
+                            it = (sr_clbk_dat->sharedMemFdList.begin() + i);
+                            if (it != sr_clbk_dat->sharedMemFdList.end()) {
+                                fdToBeClosed = sr_clbk_dat->sharedMemFdList[i].second;
+                                sr_clbk_dat->sharedMemFdList.erase(it);
+                                ALOGV("Removing fd [input %d - dup %d]", input_fd, fdToBeClosed);
+                            }
+                            break;
                         }
-                        break;
                     }
                 }
             }
         }
         PAL::getInstance()->mClientLock.unlock();
 
-        rwDonePayloadHidl.resize(sizeof(pal_callback_buffer));
-        rwDonePayload =(PalCallbackBuffer *)rwDonePayloadHidl.data();
+        rwDonePayloadHidl.resize(sizeof(struct pal_event_read_write_done_payload));
+        rwDonePayload =(PalEventReadWriteDonePayload *)rwDonePayloadHidl.data();
+        rwDonePayload->tag = rw_done_payload->tag;
         rwDonePayload->status = rw_done_payload->status;
-        switch (rw_done_payload->md_status) {
-            case ENOTRECOVERABLE: {
-                ALOGE("%s: Error, md cannot be parsed in buffer", __func__);
-                rwDonePayload->status = rw_done_payload->md_status;
-                break;
-            }
-            case EOPNOTSUPP: {
-                ALOGE("%s: Error, md id not recognized in buffer", __func__);
-                rwDonePayload->status = rw_done_payload->md_status;
-                break;
-            }
-            case ENOMEM: {
-                ALOGE("%s: Error, md buffer size received is small", __func__);
-                rwDonePayload->status = rw_done_payload->md_status;
-                break;
-            }
-            default: {
-                if (rw_done_payload->md_status) {
-                    ALOGE("%s: Error received during callback, md status = 0x%x",
-                           __func__, rw_done_payload->md_status);
-                    rwDonePayload->status = rw_done_payload->md_status;
-                }
-                break;
-            }
-        }
+        rwDonePayload->md_status = rw_done_payload->md_status;
 
-        if (!rwDonePayload->status) {
-            auto metadataParser = std::make_unique<MetadataParser>();
-            if (event_id == PAL_STREAM_CBK_EVENT_READ_DONE) {
-                auto cb_buf_info = std::make_unique<pal_clbk_buffer_info>();
-                rwDonePayload->status = metadataParser->parseMetadata(
-                            rw_done_payload->buff.metadata,
-                            rw_done_payload->buff.metadata_size,
-                            cb_buf_info.get());
-                rwDonePayload->cbBufInfo.frame_index = cb_buf_info->frame_index;
-                rwDonePayload->cbBufInfo.sample_rate = cb_buf_info->sample_rate;
-                rwDonePayload->cbBufInfo.channel_count = cb_buf_info->channel_count;
-                rwDonePayload->cbBufInfo.bit_width = cb_buf_info->bit_width;
-            } else if (event_id == PAL_STREAM_CBK_EVENT_WRITE_READY) {
-                rwDonePayload->status = getInputBufferIndex(
-                            rw_done_payload->buff.alloc_info.alloc_handle,
-                            rw_done_payload->buff.alloc_info.offset,
-                            rwDonePayload->cbBufInfo.frame_index);
-            }
-            ALOGV("%s: frame_index=%lu", __func__, rwDonePayload->cbBufInfo.frame_index);
-        }
-
-        rwDonePayload->size = rw_done_payload->buff.size;
+        rwDonePayload->buff.offset = rw_done_payload->buff.offset;
+        rwDonePayload->buff.flags = rw_done_payload->buff.flags;
+        rwDonePayload->buff.size = rw_done_payload->buff.size;
         if (rw_done_payload->buff.ts != NULL) {
-            rwDonePayload->timeStamp.tvSec = rw_done_payload->buff.ts->tv_sec;
-            rwDonePayload->timeStamp.tvNSec = rw_done_payload->buff.ts->tv_nsec;
+            rwDonePayload->buff.timeStamp.tvSec = rw_done_payload->buff.ts->tv_sec;
+            rwDonePayload->buff.timeStamp.tvNSec = rw_done_payload->buff.ts->tv_nsec;
         }
         if ((rw_done_payload->buff.buffer != NULL) &&
              !(sr_clbk_dat->session_attr.flags & PAL_STREAM_FLAG_EXTERN_MEM)) {
-            rwDonePayload->buffer.resize(rwDonePayload->size);
-            memcpy(rwDonePayload->buffer.data(), rw_done_payload->buff.buffer,
-                   rwDonePayload->size);
+            rwDonePayload->buff.buffer.resize(rwDonePayload->buff.size);
+            memcpy(rwDonePayload->buff.buffer.data(), rw_done_payload->buff.buffer,
+                   rwDonePayload->buff.size);
+        }
+        if ((rw_done_payload->buff.metadata_size > 0) &&
+             rw_done_payload->buff.metadata) {
+            ALOGV("metadatasize %zu ", rw_done_payload->buff.metadata_size);
+            rwDonePayload->buff.metadataSz = rw_done_payload->buff.metadata_size;
+            rwDonePayload->buff.metadata.resize(rwDonePayload->buff.metadataSz);
+            memcpy(rwDonePayload->buff.metadata.data(), rw_done_payload->buff.metadata,
+                    rwDonePayload->buff.metadataSz);
         }
 
+        allocHidlHandle->data[0] = rw_done_payload->buff.alloc_info.alloc_handle;
+        allocHidlHandle->data[1] = input_fd;
         ALOGV("fd [input %d - dup %d]", input_fd, rw_done_payload->buff.alloc_info.alloc_handle);
+        rwDonePayload->buff.alloc_info.alloc_handle = hidl_memory("arpal_alloc_handle",
+                                                       hidl_handle(allocHidlHandle),
+                                                       rw_done_payload->buff.alloc_info.alloc_size);
+
+        rwDonePayload->buff.alloc_info.alloc_size = rw_done_payload->buff.alloc_info.alloc_size;
+        rwDonePayload->buff.alloc_info.offset = rw_done_payload->buff.alloc_info.offset;
         if (!sr_clbk_dat->client_died) {
-            clbk_bdr->event_callback_rw_done((uint64_t)stream_handle, event_id,
-                    sizeof(pal_callback_buffer),
-                    rwDonePayloadHidl,
-                    sr_clbk_dat->client_data_);
+            auto status = clbk_bdr->event_callback_rw_done((uint64_t)stream_handle, event_id,
+                              sizeof(struct pal_event_read_write_done_payload),
+                              rwDonePayloadHidl,
+                              sr_clbk_dat->client_data_);
+            if (!status.isOk()) {
+                 ALOGE("%s: HIDL call failed during event_callback_rw_done ", __func__);
+            }
         } else
             ALOGE("Client died dropping this event %d", event_id);
 
@@ -315,6 +253,8 @@
         } else {
             ALOGE("Error finding fd %d", rw_done_payload->buff.alloc_info.alloc_handle);
         }
+        if (allocHidlHandle)
+            native_handle_delete(allocHidlHandle);
     } else {
         hidl_vec<uint8_t> PayloadHidl;
         PayloadHidl.resize(event_data_size);
@@ -691,19 +631,11 @@
 
     in_buf_cfg.buf_count = in_buff_config.buf_count;
     in_buf_cfg.buf_size = in_buff_config.buf_size;
-    if (in_buff_config.max_metadata_size) {
-        in_buf_cfg.max_metadata_size = in_buff_config.max_metadata_size;
-    } else {
-        in_buf_cfg.max_metadata_size = MetadataParser::WRITE_METADATA_MAX_SIZE();
-    }
+    in_buf_cfg.max_metadata_size =  in_buff_config.max_metadata_size;
 
     out_buf_cfg.buf_count = out_buff_config.buf_count;
     out_buf_cfg.buf_size = out_buff_config.buf_size;
-    if (out_buff_config.max_metadata_size) {
-        out_buf_cfg.max_metadata_size = out_buff_config.max_metadata_size;
-    } else {
-        out_buf_cfg.max_metadata_size = MetadataParser::READ_METADATA_MAX_SIZE();
-    }
+    out_buf_cfg.max_metadata_size =  out_buff_config.max_metadata_size;
 
     ret = pal_stream_set_buffer_size((pal_stream_handle_t *)streamHandle,
                                     &in_buf_cfg, &out_buf_cfg);
@@ -739,45 +671,41 @@
 
 Return<int32_t> PAL::ipc_pal_stream_write(const uint64_t streamHandle,
                                           const hidl_vec<PalBuffer>& buff_hidl) {
+    int32_t ret = -ENOMEM;
     struct pal_buffer buf = {0};
+    uint32_t bufSize;
+    const native_handle *allochandle = nullptr;
 
     if (!isValidstreamHandle(streamHandle)) {
         ALOGE("%s: Invalid streamHandle: %lu", __func__, streamHandle);
         return -EINVAL;
     }
 
-    buf.size = buff_hidl.data()->size;
-    std::vector<uint8_t> dataBuffer;
-    if (buff_hidl.data()->buffer.size() == buf.size) {
-        dataBuffer.resize(buf.size);
-        buf.buffer = dataBuffer.data();
-    }
+    bufSize = buff_hidl.data()->size;
+    if (buff_hidl.data()->buffer.size() == bufSize)
+        buf.buffer = (uint8_t *)calloc(1, bufSize);
+    buf.size = (size_t)bufSize;
     buf.offset = (size_t)buff_hidl.data()->offset;
-    auto timeStamp = std::make_unique<timespec>();
-    timeStamp->tv_sec =  buff_hidl.data()->timeStamp.tvSec;
-    timeStamp->tv_nsec = buff_hidl.data()->timeStamp.tvNSec;
-    buf.ts = timeStamp.get();
-    buf.flags = buff_hidl.data()->flags;
-    buf.frame_index = buff_hidl.data()->frame_index;
-
-    buf.metadata_size = MetadataParser::WRITE_METADATA_MAX_SIZE();
-    std::vector<uint8_t> bufMetadata(buf.metadata_size, 0);
-    buf.metadata = bufMetadata.data();
-    auto stream_media_config = std::make_shared<pal_media_config>();
-    for (auto& s: PAL::getInstance()->mPalClients) {
-        std::lock_guard<std::mutex> lock(s->mActiveSessionsLock);
-        for (auto session : s->mActiveSessions) {
-            if (session.session_handle == streamHandle) {
-                memcpy((uint8_t *)stream_media_config.get(),
-                       (uint8_t *)&session.callback_binder->session_attr.out_media_config,
-                       sizeof(pal_media_config));
-            }
-        }
+    buf.ts = (struct timespec *) calloc(1, sizeof(struct timespec));
+    if (!buf.ts) {
+        ALOGE("Not enough memory for buf.ts ");
+        goto exit;
     }
-    auto metadataParser = std::make_unique<MetadataParser>();
-    metadataParser->fillMetaData(buf.metadata, buf.frame_index, buf.size,
-                                 stream_media_config.get());
-    const native_handle *allochandle = buff_hidl.data()->alloc_info.alloc_handle.handle();
+    buf.ts->tv_sec =  buff_hidl.data()->timeStamp.tvSec;
+    buf.ts->tv_nsec = buff_hidl.data()->timeStamp.tvNSec;
+    buf.flags = buff_hidl.data()->flags;
+    if (buff_hidl.data()->metadataSz) {
+        buf.metadata_size = buff_hidl.data()->metadataSz;
+        buf.metadata = (uint8_t *)calloc(1, buf.metadata_size);
+        if (!buf.metadata) {
+            ALOGE("Not enough memory for buf.metadata");
+            goto exit;
+        }
+        memcpy(buf.metadata, buff_hidl.data()->metadata.data(),
+               buf.metadata_size);
+    }
+
+    allochandle = buff_hidl.data()->alloc_info.alloc_handle.handle();
 
     buf.alloc_info.alloc_handle = dup(allochandle->data[0]);
     add_input_and_dup_fd(streamHandle, allochandle->data[1], buf.alloc_info.alloc_handle);
@@ -786,33 +714,45 @@
     buf.alloc_info.alloc_size = buff_hidl.data()->alloc_info.alloc_size;
     buf.alloc_info.offset = buff_hidl.data()->alloc_info.offset;
 
+    if (buf.buffer != NULL)
+        memcpy(buf.buffer, buff_hidl.data()->buffer.data(), bufSize);
+    ALOGV("%s:%d sz %d", __func__,__LINE__,bufSize);
+    ret = pal_stream_write((pal_stream_handle_t *)streamHandle, &buf);
+exit:
     if (buf.buffer)
-        memcpy(buf.buffer, buff_hidl.data()->buffer.data(), buf.size);
-    ALOGV("%s:%d sz %zu, frame_index %lu", __func__,__LINE__, buf.size, buf.frame_index);
-
-    addToPendingInputs(buf.alloc_info.alloc_handle,
-                       buf.alloc_info.offset, buf.frame_index);
-
-    return pal_stream_write((pal_stream_handle_t *)streamHandle, &buf);
+        free(buf.buffer);
+    if (buf.ts)
+        free(buf.ts);
+    if (buf.metadata)
+        free(buf.metadata);
+    return ret;
 }
 
 Return<void> PAL::ipc_pal_stream_read(const uint64_t streamHandle,
                                       const hidl_vec<PalBuffer>& inBuff_hidl,
                                       ipc_pal_stream_read_cb _hidl_cb) {
     struct pal_buffer buf;
+    int32_t ret = 0;
     hidl_vec<PalBuffer> outBuff_hidl;
+    uint32_t bufSize;
+    const native_handle *allochandle = nullptr;
+
+    bufSize = inBuff_hidl.data()->size;
+    buf.buffer = (uint8_t *)calloc(1, bufSize);
+    buf.size = (size_t)bufSize;
+    buf.metadata_size = inBuff_hidl.data()->metadataSz;
+    buf.metadata = (uint8_t *)calloc(1, buf.metadata_size);
+    if (!buf.metadata) {
+        ALOGE("Not enough memory for buf.metadata");
+        goto exit;
+    }
 
     if (!isValidstreamHandle(streamHandle)) {
         ALOGE("%s: Invalid streamHandle: %lu", __func__, streamHandle);
         return Void();
     }
 
-    buf.size = inBuff_hidl.data()->size;
-    std::vector<uint8_t> dataBuffer(buf.size, 0);
-    buf.buffer = dataBuffer.data();
-    buf.metadata_size = MetadataParser::READ_METADATA_MAX_SIZE();
-
-    const native_handle *allochandle = inBuff_hidl.data()->alloc_info.alloc_handle.handle();
+    allochandle = inBuff_hidl.data()->alloc_info.alloc_handle.handle();
 
     buf.alloc_info.alloc_handle = dup(allochandle->data[0]);
     add_input_and_dup_fd(streamHandle, allochandle->data[1], buf.alloc_info.alloc_handle);
@@ -821,7 +761,7 @@
     buf.alloc_info.alloc_size = inBuff_hidl.data()->alloc_info.alloc_size;
     buf.alloc_info.offset = inBuff_hidl.data()->alloc_info.offset;
 
-    int32_t ret = pal_stream_read((pal_stream_handle_t *)streamHandle, &buf);
+    ret = pal_stream_read((pal_stream_handle_t *)streamHandle, &buf);
     if (ret > 0) {
         outBuff_hidl.resize(sizeof(struct pal_buffer));
         outBuff_hidl.data()->size = (uint32_t)buf.size;
@@ -833,8 +773,18 @@
           outBuff_hidl.data()->timeStamp.tvSec = buf.ts->tv_sec;
           outBuff_hidl.data()->timeStamp.tvNSec = buf.ts->tv_nsec;
         }
+        if (buf.metadata_size) {
+           outBuff_hidl.data()->metadata.resize(buf.metadata_size);
+           memcpy(outBuff_hidl.data()->metadata.data(),
+                  buf.metadata, buf.metadata_size);
+        }
     }
     _hidl_cb(ret, outBuff_hidl);
+exit:
+    if (buf.buffer)
+        free(buf.buffer);
+    if (buf.metadata)
+        free(buf.metadata);
     return Void();
 }
 
